controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java

```diff
@@ -40,13 +60,90 @@ public String getName() {
         return store.deleteNode(scopePath);
     }
 
+    @Override
+    public CompletableFuture<Void> addStreamToScope(String name, long creationTime) {
+        Integer set = findSetForStream(name);
+        String setPath = getSetPath(set);
+        String path = ZKPaths.makePath(setPath, Scope.encodeStreamInScope(name, creationTime));
+        return Futures.toVoid(store.createZNodeIfNotExist(path));
+    }
+
+    @Override
+    public CompletableFuture<Void> removeStreamFromScope(String name, long creationTime) {
+        Integer set = findSetForStream(name);
+        String setPath = getSetPath(set);
+        String path = ZKPaths.makePath(setPath, Scope.encodeStreamInScope(name, creationTime));
+        return Futures.toVoid(store.deletePath(path, true));
+    }
+
     @Override
     public CompletableFuture<List<String>> listStreamsInScope() {
         return store.getChildren(scopePath);
     }
 
     @Override
-    public void refresh() {
+    public CompletableFuture<Pair<List<String>, String>> listStreamsInScope(int limit, String continuationToken, Executor executor) {
+        AtomicInteger set = new AtomicInteger(0);
+        AtomicReference<String> floor = new AtomicReference<>();
+        if (!Strings.isNullOrEmpty(continuationToken)) {
+            Pair<String, Long> decoded = Scope.decodeStreamInScope(continuationToken);
```

Comment:
If this continuation token is external, it needs to be versioned, Please see my comment in `Scope.decodeStreamInScope`

Reply:
It has been made versioned in branch #3312